File: Emulator/src/main/core/cache/def/impl/AnimationDefinition.java
package core.cache.def.impl;

import core.cache.Cache;
import core.cache.CacheIndex;
import core.cache.misc.buffer.ByteBufferUtils;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;

/**
 * The type Animation definition.
 */
public final class AnimationDefinition {
    /**
     * The An int 2136.
     */
    public int anInt2136;
    /**
     * The An int 2137.
     */
    public int anInt2137;
    /**
     * The An int array 2139.
     */
    public int[] anIntArray2139;
    /**
     * The An int 2140.
     */
    public int anInt2140;
    /**
     * The Has sound effect.
     */
    public boolean hasSoundEffect = false;
    /**
     * The An int 2142.
     */
    public int anInt2142;
    /**
     * The Emote item.
     */
    public int emoteItem;
    /**
     * The An int 2144.
     */
    public int anInt2144 = -1;
    /**
     * The Handled sounds.
     */
    public int[][] handledSounds;
    /**
     * The A boolean array 2149.
     */
    public boolean[] aBooleanArray2149;
    /**
     * The An int array 2151.
     */
    public int[] anIntArray2151;
    /**
     * The A boolean 2152.
     */
    public boolean aBoolean2152;
    /**
     * The Durations.
     */
    public int[] durations;
    /**
     * The An int 2155.
     */
    public int anInt2155;
    /**
     * The A boolean 2158.
     */
    public boolean aBoolean2158;
    /**
     * The A boolean 2159.
     */
    public boolean aBoolean2159;
    /**
     * The An int 2162.
     */
    public int anInt2162;
    /**
     * The An int 2163.
     */
    public int anInt2163;
    /**
     * The New header.
     */
    boolean newHeader;

    /**
     * The Sound min delay.
     */
    public int[] soundMinDelay;
    /**
     * The Sound max delay.
     */
    public int[] soundMaxDelay;
    /**
     * The An int array 1362.
     */
    public int[] anIntArray1362;
    /**
     * The Effect 2 sound.
     */
    public boolean effect2Sound;

	private static final Map<Integer, AnimationDefinition> animationDefinition = new HashMap<>();

    /**
     * For id animation definition.
     *
     * @param emoteId the emote id
     * @return the animation definition
     */
    public static final AnimationDefinition forId(int emoteId) {
		try {
			AnimationDefinition defs = animationDefinition.get(emoteId);
			if (defs != null) {
				return defs;
			}
			byte[] data = Cache.getData(CacheIndex.SEQUENCE_CONFIGURATION, emoteId >>> 7, emoteId & 0x7f);
			defs = new AnimationDefinition();
			if (data != null) {
				defs.readValueLoop(ByteBuffer.wrap(data));
			}
			defs.method2394();
			animationDefinition.put(emoteId, defs);
			return defs;
		} catch (Throwable t) {
			return null;
		}
	}

	private void readValueLoop(ByteBuffer buffer) {
		for (;;) {
			int opcode = buffer.get() & 0xFF;
			if (opcode == 0) {
				break;
			}
			readValues(buffer, opcode);
		}
	}

    /**
     * Gets duration.
     *
     * @return the duration
     */
    public int getDuration() {
		if (durations == null) {
			return 0;
		}
		int duration = 0;
		for (int i : durations) {
			if (i > 100) {
				continue;
			}
			duration += i * 20;
		}
		return duration;
	}

    /**
     * Gets cycles.
     *
     * @return the cycles
     */
    public int getCycles() {
		if (durations == null) return 0;
		int duration = 0;
		for (int i : durations)
			duration += i;
		return duration;
	}

    /**
     * Gets duration ticks.
     *
     * @return the duration ticks
     */
    public int getDurationTicks() {
		int ticks = getDuration() / 600;
		return Math.max(ticks, 1);
	}

	private void readValues(ByteBuffer buffer, int opcode) {
		if (opcode == 1) {
			int length = buffer.getShort() & 0xFFFF;
			durations = new int[length];
			for (int i = 0; i < length; i++) {
				durations[i] = buffer.getShort() & 0xFFFF;
			}
			anIntArray2139 = new int[length];
			for (int i = 0; i < length; i++) {
				anIntArray2139[i] = buffer.getShort() & 0xFFFF;
			}
			for (int i = 0; i < length; i++) {
				anIntArray2139[i] = ((buffer.getShort() & 0xFFFF << 16) + anIntArray2139[i]);
			}
		} else if (opcode != 2) {
			if (opcode != 3) {
				if (opcode == 4)
					aBoolean2152 = true;
				else if (opcode == 5)
					anInt2142 = buffer.get() & 0xFF;
				else if (opcode != 6) {
					if (opcode == 7)
						emoteItem = buffer.getShort() & 0xFFFF;
					else if ((opcode ^ 0xffffffff) != -9) {
						if (opcode != 9) {
							if (opcode != 10) {
								if (opcode == 11)
									anInt2155 = buffer.get() & 0xFF;
								else if (opcode == 12) {
									int i = buffer.get() & 0xFF;
									anIntArray2151 = new int[i];
									for (int i_19_ = 0; ((i_19_ ^ 0xffffffff) > (i ^ 0xffffffff)); i_19_++) {
										anIntArray2151[i_19_] = buffer.getShort() & 0xFFFF;
									}
									for (int i_20_ = 0; i > i_20_; i_20_++)
										anIntArray2151[i_20_] = ((buffer.getShort() & 0xFFFF << 16) + anIntArray2151[i_20_]);
								} else if (opcode == 13) {
									// opcode 13
									int i = buffer.getShort() & 0xFFFF;
									handledSounds = new int[i][];
									for (int i_21_ = 0; i_21_ < i; i_21_++) {
										int i_22_ = buffer.get() & 0xFF;
										if ((i_22_ ^ 0xffffffff) < -1) {
											handledSounds[i_21_] = new int[i_22_];
											handledSounds[i_21_][0] = ByteBufferUtils.getMedium(buffer);
											for (int i_23_ = 1; ((i_22_ ^ 0xffffffff) < (i_23_ ^ 0xffffffff)); i_23_++) {
												handledSounds[i_21_][i_23_] = buffer.getShort() & 0xFFFF;
											}
										}
									}
								} else if (opcode == 14) {
									hasSoundEffect = true;
								} else {

								}
							} else
								anInt2162 = buffer.get() & 0xFF;
						} else
							anInt2140 = buffer.get() & 0xFF;
					} else
						anInt2136 = buffer.get() & 0xFF;
				} else
					anInt2144 = buffer.getShort() & 0xFFFF;
			} else {
				aBooleanArray2149 = new boolean[256];
				int length = buffer.get() & 0xFF;
				for (int i = 0; i < length; i++) {
					aBooleanArray2149[buffer.get() & 0xFF] = true;
				}
			}
		} else
			anInt2163 = buffer.getShort() & 0xFFFF;
	}

    /**
     * Method 2394.
     */
    public void method2394() {
		if (anInt2140 == -1) {
			if (aBooleanArray2149 == null)
				anInt2140 = 0;
			else
				anInt2140 = 2;
		}
		if (anInt2162 == -1) {
			if (aBooleanArray2149 == null)
				anInt2162 = 0;
			else
				anInt2162 = 2;
		}
	}

    /**
     * Instantiates a new Animation definition.
     */
    public AnimationDefinition() {
		anInt2136 = 99;
		emoteItem = -1;
		anInt2140 = -1;
		aBoolean2152 = false;
		anInt2142 = 5;
		aBoolean2159 = false;
		anInt2163 = -1;
		anInt2155 = 2;
		aBoolean2158 = false;
		anInt2162 = -1;
	}

    /**
     * Gets definition.
     *
     * @return the definition
     */
    public static Map<Integer, AnimationDefinition> getDefinition() {
		return animationDefinition;
	}
}

File: Emulator/src/main/core/cache/def/impl/AnimationDefinition.kt
package core.cache.def.impl

import core.cache.Cache
import core.cache.CacheIndex
import core.cache.misc.buffer.ByteBufferUtils
import java.nio.ByteBuffer

/**
 * Represents an animation definition.
 */
class AnimationDefinition {
    var maxLoops = 99
    var movementPriority = 0
    var frames: IntArray? = null
    var priorityOverride = -1
    var hasSoundEffect = false
    var priority = 5
    var rightHandItem = -1
    var leftHandItem = -1
    var handledSounds: Array<IntArray?>? = null
    var interleaveOrder: BooleanArray? = null
    var expressionFrames: IntArray? = null
    var isForcedPriority = false
    var duration: IntArray? = null
    var replayMode = 2
    var disableResetOnLoop = false
    var stopsOnMovement = false
    var priorityBackup = -1
    var loopOffset = -1
    var newHeader = false
    var soundMinDelay: IntArray? = null
    var soundMaxDelay: IntArray? = null
    var frameSoundEffects: IntArray? = null
    var effect2Sound = false

    companion object {
        private val animationDefinition = mutableMapOf<Int, AnimationDefinition>()

        /**
         * Retrieves an [AnimationDefinition] by its emote id.
         *
         * @param emoteId The id of the animation.
         * @return The [AnimationDefinition] if found, or `null` otherwise.
         */
        @JvmStatic
        fun forId(emoteId: Int): AnimationDefinition? =
            try {
                animationDefinition[emoteId] ?: run {
                    val data = Cache.getData(CacheIndex.SEQUENCE_CONFIGURATION, emoteId ushr 7, emoteId and 0x7f)
                    val definition = AnimationDefinition()
                    data?.let { definition.readValueLoop(ByteBuffer.wrap(it)) }
                    definition.changeValues()
                    animationDefinition[emoteId] = definition
                    definition
                }
            } catch (_: Throwable) {
                null
            }

        /**
         * Retrieves all loaded animation definitions.
         *
         * @return A map containing animation definitions by their id.
         */
        @JvmStatic
        fun getDefinition(): Map<Int, AnimationDefinition> = animationDefinition
    }

    /**
     * Reads animation values from a buffer.
     *
     * @param buffer The [ByteBuffer] containing animation data.
     */
    private fun readValueLoop(buffer: ByteBuffer) {
        while (true) {
            val opcode = buffer.get().toInt() and 0xFF
            if (opcode == 0) break
            readValues(buffer, opcode)
        }
    }

    /**
     * Gets the total duration of the animation in milliseconds.
     *
     * @return The animation duration in milliseconds.
     */
    fun getDuration(): Int = duration?.filter { it <= 100 }?.sumOf { it * 20 } ?: 0

    /**
     * Gets the total cycle count for the animation.
     *
     * @return The number of cycles.
     */
    fun getCycles(): Int = duration?.sum() ?: 0

    /**
     * Gets the duration of the animation in game ticks.
     *
     * @return The duration in game ticks.
     */
    fun getDurationTicks(): Int = maxOf(getDuration() / 600, 1)

    /**
     * Reads values from a buffer based on opcodes.
     *
     * @param buffer The [ByteBuffer] containing the data.
     * @param opcode The opcode indicating what data to read.
     */
    private fun readValues(
        buffer: ByteBuffer,
        opcode: Int,
    ) {
        when (opcode) {
            1 -> {
                val length = buffer.short.toInt() and 0xFFFF
                duration = IntArray(length) { buffer.short.toInt() and 0xFFFF }
                frames = IntArray(length) { buffer.short.toInt() and 0xFFFF }
                frames?.forEachIndexed { i, v ->
                    frames!![i] = (buffer.short.toInt() and 0xFFFF shl 16) + v
                }
            }

            2 -> loopOffset = buffer.short.toInt() and 0xFFFF
            3 -> {
                interleaveOrder = BooleanArray(256)
                val length = buffer.get().toInt() and 0xFF
                repeat(length) {
                    interleaveOrder!![buffer.get().toInt() and 0xFF] = true
                }
            }

            4 -> isForcedPriority = true
            5 -> priority = buffer.get().toInt() and 0xFF
            6 -> leftHandItem = buffer.short.toInt() and 0xFFFF
            7 -> rightHandItem = buffer.short.toInt() and 0xFFFF
            8 -> maxLoops = buffer.get().toInt() and 0xFF
            9 -> priorityOverride = buffer.get().toInt() and 0xFF
            10 -> priorityBackup = buffer.get().toInt() and 0xFF
            11 -> replayMode = buffer.get().toInt() and 0xFF
            12 -> {
                val i = buffer.get().toInt() and 0xFF
                expressionFrames = IntArray(i) { buffer.short.toInt() and 0xFFFF }
                expressionFrames?.forEachIndexed { index, v ->
                    expressionFrames!![index] = (buffer.short.toInt() and 0xFFFF shl 16) + v
                }
            }

            13 -> {
                val i = buffer.short.toInt() and 0xFFFF
                handledSounds = Array(i) { null }
                repeat(i) { index ->
                    val size = buffer.get().toInt() and 0xFF
                    if (size > 0) {
                        handledSounds!![index] =
                            IntArray(size).apply {
                                this[0] = ByteBufferUtils.getMedium(buffer)
                                for (j in 1 until size) this[j] = buffer.short.toInt() and 0xFFFF
                            }
                    }
                }
            }

            14 -> hasSoundEffect = true
        }
    }

    /**
     * Adjusts default values.
     */
    fun changeValues() {
        if (priorityOverride == -1) {
            priorityOverride = if (interleaveOrder == null) 0 else 2
        }
        if (priorityBackup == -1) {
            priorityBackup = if (interleaveOrder == null) 0 else 2
        }
    }
}
